tecken/libsentry.py

```diff
@@ -0,0 +1,400 @@
+            return
+
+        elif part in parent:
+            parent = parent[part]
+
+    if isinstance(parent, dict):
+        yield parent
+
+
+class Scrubber:
+    """Scrubber pipeline for Sentry events
+
+    https://docs.sentry.io/platforms/python/configuration/filtering/
+
+    """
+
+    def __init__(self, scrub_keys=SCRUB_KEYS_DEFAULT):
+        """
+        :arg scrub_keys: list of ``(key_path, keys, scrub function)`` tuples
+
+            ``key_path`` is a Python dotted path of key names with ``[]`` to denote
+            arrays to traverse pointing to a dict with values to scrub.
+
+            ``keys`` is a list of keys to scrub values of
+
+            A scrub function takes a value and returns a scrubbed value. For
+            example::
+
+                def hide_letter_a(value):
+                    return "".join([letter if letter != "a" else "*" for letter in value])
+
+            Example of ``scrub_keys``::
+
+                ("request.data", ("csrfmiddlewaretoken",), scrub()),
+
+                ("exception.stacktrace.frames.[].vars", ("code_id",), scrub()),
+
+        """
+        # Split key_path into parts and verify that scrub_keys has the right shape
+        self.scrub_keys = [
+            (key_path.split("."), keys, scrub_function)
+            for key_path, keys, scrub_function in scrub_keys
+        ]
+
+    def __call__(self, event, hint):
+        """Implements before_send function interface and scrubs Sentry event
+
+        This tries really hard to be very defensive such that even if there are bugs in
+        the scrubs, it still emits something to Sentry.
+
+        It will log errors, so we should look for those log statements. They'll all have
+        "LIBSENTRYERROR" in the message making them easy to find regardless of the
+        logger name.
+
+        Further, they emit two incr metrics:
+
+        * scrub_fun_error
+        * get_target_dicts_error
+
+        Put those in a dashboard with alerts so you know when to look in the logs.
+
+        """
+
+        for key_path, keys, scrub_fun in self.scrub_keys:
+            try:
+                for parent in get_target_dicts(event, key_path):
+                    if not parent:
+                        continue
+
+                    for key in keys:
+                        if key not in parent:
+                            continue
+
+                        val = parent[key]
+
+                        try:
+                            filtered_val = scrub_fun(val)
+                        except Exception:
+                            logger.exception(f"LIBSENTRYERROR: Error in {scrub_fun}")
+                            metrics.incr("scrub_fun_error")
+                            filtered_val = "ERROR WHEN SCRUBBING"
+
+                        parent[key] = filtered_val
+            except Exception:
+                logger.exception("LIBSENTRYERROR: Error in get_target_dicts")
+                metrics.incr("get_target_dicts_error")
```

Comment:
If we wanted to make this into a library we published on PyPI, we'd remove markus usage here (not everyone uses markus) and replace it with a hook for reporting an error with the scrubber.